File: app/javascript/channels/activity_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("ActivityChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    this.perform("appear");
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    // let statusDots = document.getElementsByClassName(`user-${data.user_id}-status`);

    // for (let i = 0; i < statusDots.length; i++) {
    //   let element = statusDots[i];
    //   if (data.status === "online") {
    //     element.classList.add("online");
    //   } else {
    //     element.classList.remove("online");
    //   }
    // }

    let statusDots = document.getElementsByClassName(`user-${data.user_id}-status`);

    for (let i = 0; i < statusDots.length; i++) {
      let element = statusDots[i];

      if (data.status === "online") {
        element.classList.add("online");
        element.classList.remove("offline");
      } else {
        element.classList.remove("online");
        element.classList.add("offline");
      }
    }
  }
});

File: app/channels/activity_channel.rb
class ActivityChannel < ApplicationCable::Channel
  def subscribed
    stream_from "activity_channel"
  end

  def unsubscribed
    online_status = OnlineStatus.find_by(user_id: current_user.id)

    if online_status.nil?
      online_status = OnlineStatus.create(
        user_id: current_user.id, 
        status: "offline"
      )
    else
      online_status.status = "offline"
      online_status.save
    end

    ActionCable.server.broadcast("activity_channel", {user_id: current_user.id, status: "offline"})
  end

  def appear
    online_status = OnlineStatus.find_by(user_id: current_user.id)

    if online_status.nil?
      online_status = OnlineStatus.create(
        user_id: current_user.id, 
        status: "online"
      )
    else
      online_status.status = "online"
      online_status.save
    end

    ActionCable.server.broadcast("activity_channel", {user_id: current_user.id, status: "online"})
  end
end
